teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -84,97 +84,399 @@ that can be shared between them, and running the second Task will be delayed by
+          value: "Tests completed with $(tasks.run-unit-tests.status) status"
+```
+
+We could leverage
+[TEP-0056 pipelines in pipelines](https://github.com/tektoncd/community/blob/main/teps/0056-pipelines-in-pipelines.md):
+to run this Pipeline within another as a pod, for example:
+
+
+```yaml
+kind: Pipeline
+metadata:
+  name: build-test-deploy
+spec:
+  workspaces:
+    - name: source-code
+  tasks:
+    - name: get-source
+      workspaces:
+        - name: source-code
+          workspace: source-code
+      taskRef:
+        name: git-clone
+      params:
+        - name: url
+            value: $(params.url)
+        - name: revision
+          value: $(params.revision)
+    - name: run-unit-tests-and-upload-results
+      runAfter: get-source
+      pipelineRef: # syntax may look different as TEP-0056 but the idea is that this is a Pipeline (in a pipeline) instead of a Task or Run
+        name: unit-tests
+      execution/scheduling: "co-located" # this pipeline in a pipeline should run in a pod
+      workspaces:
+        - name: source-code
+          workspcae: source-code
+    - name: deploy
+      runAfter: [run-unit-tests-and-upload-results]
+      when:
+        - input: "$(params.git-branch)"
+          operator: in
+          values: ["main"]
+      taskRef:
+        name: build-deploy
+      workspaces:
+        - name: source-code
+          workspcae: source-code
+```
+
+In the above Pipeline, the Task `git-clone` would run as a pod, the Pipeline `unit-test` would be executed as a pod
+(by creating a PipelineRun using the new syntax) - including two `finally` Tasks, and the Task `build-deploy` would run
+as a pod.
 
+The Pipeline `unit-test` also takes a workspace `test-results` but when executing as a pod, this can be satisfied by
+`emptyDir`, so in the Pipeline above that runs `unit-test` as a Pipeline in a pod, it skips specifying this and since
+the execution mode is `co-located` the PipelineRun can be constructed to pass `emptyDir` (or that could be the
+implied default for any workspaces when a PipelineRun is run in `co-located` mode).
 
-### Task composition in Pipeline Tasks
+Pros:
+* Doesn't involve introduction of new types
+* Assuming
+  [TEP-0056 pipelines in pipelines](https://github.com/tektoncd/community/blob/main/teps/0056-pipelines-in-pipelines.md)
+  users can construct composition of Pipelines however they'd like (i.e. parts of a Pipeline in a pod, parts not)
+Cons:
+* A Pipeline run in "co-located" mode could itself contain Pipelines to run in "co-located" mode which might be a bit
+  weird (and this nesting could continue.... infinitely?)
+  * Ultimately this would mean the entire flattened Pipeline would run in a pod so it shouldn't be ambiguous
+  * Probably makes sense for Pipeline in Pipeline to have some kind of depth limit defined anyway?
 
-In this option we make it possible to express Tasks which can be combined together to run sequentially as one pod.
+####### init/before finally/after
```

Comment:
I think those would be on the `Task` (or `TaskRun`) iteslf, no on `Pipeline` objects (a bit similar to https://github.com/tektoncd/community/pull/502 or https://github.com/tektoncd/community/pull/369). The idea then would be : 
- They obey the `Task` law (all in sequence)
- This is explicit, the user choose what to run in the same pod or not
- They would allow things such as a single `TaskRun` to build a image based of some source (by *injecting* a git-clone task in there)